Shields: Allow shields conditional configuration

erwango opened the pull request:
Today shields Kconfig configuration (set of Kconfig symbols defined in `<shield>_defconfig` files) is applied unconditionally when a specific shield is selected by user (or indirectly in a sample/test via cmake configuration or with a yaml configuration when running sanitycheck).

This must be changed. Indeed, for application portability matter, board default configuration guidelines states that features activation is application responsibility and should not be done on board side (and shields are on board side).
This is applied for boards and should then be aslo applied for shields, so applications can rely on the fact that a HW component can be configured the same way, either it is directly available on the board or added via a shield.

To make shield conditional configuration possible:
- Get cmake variable SHIELD_AS_LIST (user selection of shields) available in Kconfig
- Set a SHIELD_MY_SHIELDX symbol, matching the state of user activation of my_shieldx in `boards/shields/my_shieldx/Kconfig.shield`
- Parse newly created `Kconfig.shield` in boards/Kconfig (which by the way enable the selection of shields in menuconfig)
- Set Kconfig shield configuration in Kconfig.defconfig under feature condition
- Parse shields/boards/*.defconfig to get boards secific shields configuration scaling in this new scheme
- Convert shields `x_nucleo_idb05a1` and `wnc_m14a2a` to this new scheme. Eventually all shields will be converted as part of this PR if RFC get positive feedback.

EDIT: 11-27-19: Pushed a new version using a more readable Kconfig function (`shields_list_contains`)

EDIT: 12-10-19: Pushed a new version which updates all shields

TODO:
- [x] Complete shields convertion
- [x] Update shields documentation

Fixes #17310

erwango committed:
script/Kconfig: kconfig functions: Add shields_list_contains function

Add shields_list_contains Kconfig function which return bool based on
check of shield presence in cmake SHIELD_AS_LIST.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
cmake: make shield list available to Kconfig

For application portability, it is required that feature activation
is made conditional in shield configuration. This way features remain
controlled on application side.
To enable this we need that list of user activated shield is made
available to Kconfig.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Kconfig: Enable use of shield conditional Kconfig files

To make use of shields Kconfig.defconfig, parse these files
and newly created Kconfig.shield files which will define the
SHIELD_XXX Kconfig symbols that will be used for conditional
configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: x_nucleo_idb05a1: Move to conditional Kconfig

Put shield configuration Kconfig flags in new Kconfig.defconfig.
This way shield BLE susbystem can now be selected by application
using BT symbol, similarly to what is done when BlueNRG ship is
directly present on board.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on Kconfig.zephyr:
```diff
@@ -22,6 +22,8 @@ endmenu
 # Note: $ARCH and $BOARD_DIR might be glob patterns.
 source "$(BOARD_DIR)/Kconfig.defconfig"
 
+source "boards/shields/*/Kconfig.defconfig"
```
> Are the enabled shields known pre-Kconfig?

Yes and no.
When running a west build command this is known... But user can modify configuration using `menuconfig`. 

> If so we want to limit the sourced files to just those enabled.

If we do so (need to figure it out how, though), we prevent the ability to select shields in `menuconfig`.
So we need to chose between this and build performance (which, I guess, is your point).

SebastianBoe replied:
> But user can modify shields selection using menuconfig.

Never mind then, thank you for the explanation.

dbkinder commented on doc/guides/kconfig/index.rst:
```diff
@@ -848,15 +848,17 @@ Kconfig Functions
+Kconfig.
 For example, we can get the default value for a Kconfig symbol
 from the devicetree.
+See the Python docstrings in ``scripts/kconfig/kconfigfunctions.py`` for more
```
OK for now, but we can also use the sphinx autodoc capabilities (as done for west).

dbkinder reviewed (approved):
OK on the docs, but we should take advantage of the sphinx autodoc capabilities (as done with west), to extract and process docstrings.

ulfalizer commented on doc/guides/kconfig/index.rst:
```diff
@@ -848,15 +848,17 @@ Kconfig Functions
 For example, we can get the default value for a Kconfig symbol
 from the devicetree.
+See the Python docstrings in ``scripts/kconfig/kconfigfunctions.py`` for more
+details on the functions.
```
Could reflow this paragraphs maybe (`gq` in Vim is handy, if you use that).

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,6 +335,15 @@ def dt_compat_enabled(kconf, _, compat):
+    Returns "y" if 'elem' appears in 'list' after split()ing it, and "n"
+    otherwise.
+    """
+
```
No blank line for consistency with other functions.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,6 +335,15 @@ def dt_compat_enabled(kconf, _, compat):
     return "n"
 
 
+def cmake_in_list(kconf, _, elem, list):
```
Could call it `in_cmake_list()` maybe. I read it as `cmake_in_list` as "CMake is in the list" at first and was a bit confused.

erwango replied:
I though about it, but finally opted to use cmake as namespace to follow logic used for other dt functions (`dt_foo`, ..).
Up to you!

mbolivar replied:
How about `cmake_list_contains`?

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,6 +335,15 @@ def dt_compat_enabled(kconf, _, compat):
+def cmake_in_list(kconf, _, elem, list):
+    """
+    Returns "y" if 'elem' appears in 'list' after split()ing it, and "n"
+    otherwise.
```
Update docstring to explain where CMake comes in.

ulfalizer commented on boards/shields/wnc_m14a2a/boards/nrf52840_pca10056.defconfig:
```diff
@@ -0,0 +1,15 @@
+if BOARD_NRF52840_PCA10056
+
+config GPIO_NRF_P1
+	def_bool y
```
The note at the end of https://docs.zephyrproject.org/latest/guides/kconfig/index.html#common-shorthands might apply here:

    For a symbol defined in multiple locations (e.g., in a Kconfig.defconfig
    file in Zephyr), it is best to only give the symbol type for the “base”
    definition of the symbol, and to use default (instead of def_<type> value)
    for the remaining definitions. That way, if the base definition of the
    symbol is removed, the symbol ends up without a type, which generates a
    warning that points to the other definitions. That makes the extra
    definitions easier to discover and remove.

Use `default` instead of `def_bool` if these already have a base definition.

ulfalizer reviewed (changes requested):
For `def_bool` stuff.

erwango commented:
@ulfalizer, Thanks for the review. I've pushed a new version fixing the points you reported. About: 

> Could call it in_cmake_list() maybe. I read cmake_in_list as "CMake is in the list" at first and was a bit confused.

I answered, let me know what you prefer.

mbolivar reviewed (approved):
Looks like a much better way to do this. +1 from me for the approach, nice work!

erwango commented:
@mbolivar, @ulfalizer, thanks for reviews. For the remaining controversial point, what about `cmake_shields_list_contains` (less generic, but clearer and cleaner) ? So in `Kconfig.shields` files we'll use :
```
config SHIELD_WNC_M14A2A
	bool "WNC-M14A2A LTE-M Modem"
	default $(cmake_shields_list_contains,wnc_m14a2a)
```
We can even drop the cmake and use:
```
	default $(shields_list_contains,wnc_m14a2a)
```

mbolivar commented:
Either `cmake_shields_list_contains` or `shields_list_contains` sounds good to me.

erwango committed:
boards/shields: Move wnc_m14a2a to conditional Kconfig

Put shield configuration Kconfig flags in new Kconfig.defconfig.
This way shield BLE subsystem can now be selected by application
using NETWORKING symbol, similarly to what is done when modem is
directly present on board.
Additionally, move board Kconfig files to a similar scheme where
shields related Kconfig flags are made available under shield,
board and NETWORKING symbols.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ulfalizer commented on boards/shields/x_nucleo_idb05a1/Kconfig.shield:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_X_NUCLEO_IDB05A
+	bool "ST BLE shield with BlueNRG"
```
Why can shields be enabled/disabled in Kconfig independently from the `SHIELD_AS_LIST` setting?

ulfalizer commented on doc/guides/kconfig/index.rst:
```diff
@@ -909,6 +909,12 @@ the above to look like:
    config FLASH_BASE_ADDRESS
       default 0x20010000
 
+Cmake related Functions
```
*CMake-related

ulfalizer commented on doc/guides/kconfig/index.rst:
```diff
@@ -909,6 +909,12 @@ the above to look like:
+
+.. code-block:: none
+
+  cmake_in_list(kconf, _, elem, list):
```
Indent by three spaces, extra `:` at the end.

erwango replied:
Done, but no more aligned with other "Devicetree Related Functions"

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,6 +335,19 @@ def dt_compat_enabled(kconf, _, compat):
+def shields_list_contains(kconf, _, shield):
+    """
+    Returns "y" if cmake environment variable 'SHIELD_AS_LIST' exists and
+    'shield' is in after split()ing it, and "n" otherwise.
```
`split()` splits on whitespace.

erwango replied:
I didn't get this comment..

ulfalizer replied:
`split()`ing would split on whitespace rather than semicolons, so the docstring is wrong.

Could say something like `Return "y" if 'shield' appears in the comma-separated list in ... Returns "n" if ...`.

erwango replied:
@ulfalizer, gotcha.

ulfalizer commented on doc/guides/kconfig/index.rst:
```diff
@@ -848,15 +848,15 @@ Kconfig Functions
-we use in Zephyr to expose devicetree information to Kconfig.
-For example, we can get the default value for a Kconfig symbol
-from the devicetree.
+we use in Zephyr to expose devicetree or cmake environment information to
```
*CMake

ulfalizer commented on boards/shields/wnc_m14a2a/Kconfig.shield:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_WNC_M14A2A
+	bool "WNC-M14A2A LTE-M Modem"
```
Why can shields be enabled/disabled in Kconfig independently from the SHIELD_AS_LIST setting?

erwango replied:
This is actually directly dependent. Though (if I understand your point), we need this SHIELD_FOO symbol to enable usage of Kconfig.defconfig files.

erwango replied:
Though, it is right that, since shield enabling depends on SHIELD_AS_LIST, there is no way we can enable it from menuconfig (we can, but it has actually no effect) contrary to what I answered to @SebastianBoe in https://github.com/zephyrproject-rtos/zephyr/pull/20934#discussion_r349540628.
So I'm gonna review this part.

ulfalizer replied:
Yeah, it's a good idea to always imagine someone going into `menuconfig` and trying really to create a broken configuration when making Kconfig changes.

erwango replied:
ok, removed the "menu" stuff

ulfalizer commented:
https://docs.zephyrproject.org/latest/guides/porting/shields.html doesn't seem to mention that there can be multiple shields.

erwango commented:
@ulfalizer 

> https://docs.zephyrproject.org/latest/guides/porting/shields.html doesn't seem to mention that there can be multiple shields.

Indeed, this should be updated.

erwango commented:
@SebastianBoe, my bad. @ulfalizer made the point that if selecting shields in `menuconfig` is possible it is actually not sufficient to enable them, because to get it fully functional it requires to update SHIELD_AS_LIST, which is not possible (yet at least).

So, to answer you comment (https://github.com/zephyrproject-rtos/zephyr/pull/20934#discussion_r349536543) I see 3 options:

a. Keep shield selection is not possible by `menuconfig` and stay with the sub-optimised search across all available shields

b. Keep shield selection is not possible by `menuconfig` and limit sourced files to just those enabled > I guess it requires an update to `kconfiglib.py`. Maybe @ulfalizer can comment on feasibility.

c. Make shield selection actually possible through `menuconfig`, which may require to rework the full CMake SHIELD handling. You may be the right one to comment here.


Since b. and c. may require deeper analysis and current PR as a base. I propose we can get current PR merged (once other comments are fixed approved and all shields have been moved to this format) and raise a dedicated issue to discuss the next direction (b. or c.).
@ulfalizer, @SebastianBoe, is that fine for you?

SebastianBoe commented:
Is it possible to disable a shield that is in the SHIELD_AS_LIST ?

erwango commented:
@SebastianBoe 

> Is it possible to disable a shield that is in the SHIELD_AS_LIST ?

It is indeed to disable the Kconfig symbol SHIELD_FOO of a foo shield present in SHIELD_AS_LIST (but it doesn't remove it from the list)

SebastianBoe commented:
Then doing "a." for now and "b." when we figure out a clean way to prune the other shields is the preferred approach.

erwango committed:
doc/guides: shields: Document possibility to enable several shields

It is actually possible to use several shields in the same project.
Reflect it in the shields documentation.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
> Then doing "a." for now and "b." when we figure out a clean way to prune the other shields is the preferred approach.

Fine for me. Thanks @SebastianBoe 

@ulfalizer, I've pushed new version of the PR fixing your comments (minus one that I'd like you to clarify). Also can you give you opinion on above proposal?

ulfalizer commented:
Is there any point in putting a shield in `SHIELD` if you don't want it enabled? I imagined `SHIELD` as a list of shields to enable support for.

ulfalizer commented:
Looks like it's still possible to turn off shields in `menuconfig` (the symbols have a prompt). Nice if we can avoid that, so that there's just a single mechanism for enabling shields.

erwango commented:
> Looks like it's still possible to turn off shields in menuconfig

@ulfalizer, yeah I was focusing on the `menu` stuff an forgot that prompt was sufficient to appear in `menuconfig`. This is fixed now, thanks.

ulfalizer commented on boards/shields/wnc_m14a2a/Kconfig.shield:
```diff
@@ -0,0 +1,6 @@
+
+config SHIELD_WNC_M14A2A
+	bool
+	default $(shields_list_contains,wnc_m14a2a)
```
Could also use `def_bool` here. It's pretty common for non-configurable "helper" symbols like this one.

```
config SHIELD_WNC_M14A2A
	def_bool $(shields_list_contains,wnc_m14a2a)
```

erwango replied:
Done, txs!

erwango committed:
boards/shields: x_nucleo_iks01ax: move to conditional Kconfig

Add Kconfig.shield file for x_nucleo_iks01ax shields.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: waveshare_epaper: Move to conditional Kconfig

Move wavehare epaper shields to conditional Kconfig.
As part of this change, direct drivers Kconfig symbols enabling
(SPI, SSD16XX) are removed as they are application
responsibility. GPIO is removed as well as it should be 'y' by
default on  all boards according to Default board configuration
guidelines.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: ssd1306: Factorize and move to conditional Kconfig

Factorize definitions for ssd1306_128x32 and sd1306_128x64 shields
and move to conditional Kconfig.
As part of this change, direct drivers Kconfig symbols enabling
(I2C, SSD1306) are removed as they are application responsibility.
Also disabling SSD16XX is removed as SSD16XX should not be enabled
by default.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: sparkfun_sara_r4: Move to conditional Kconfig

Move sparkfun_sara_r4 to conditional Kconfig.
As part of this change, remove disco_l475_iot1 specific configuration
as it is already part of board default configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: link_board_eth: Move to conditional Kconfig

Move link_board_eth to conditional Kconfig.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: link_board_can: Move to conditional Kconfig

Move link_board_can to conditional Kconfig.
As part of this change, some board specific symbols are moved
to reel_board specific file.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: frdm_kw41z: Move to conditional Kconfig

Add Kconfig.shield to frdm_kw41z.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: fdrm_cr20a: Move to conditional Kconfig

Move shield to Kconfig.shield format.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: dfrobot_can_bus_v2_0: Move to Kconfig.shield format

Move dfrobot_can_bus_v2_0 to Kcondig.shield foramt and as part of
this change introduce nrf52_pca10040 board specific configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards/shields: adafruit_2_8_tft_touch_v2: Move to Kconfig.defconfig

Move adafruit_2_8_tft_touch_v2 to Kconfig.defconfig format.
As part of this change Kconfig flags SPI and DISPLAY are removed
from the shield configuration as they are part of application
configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
doc/guides: shields: Document shield conditional configuration

Update doc with new Kconfig.shield files.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mike-scott reviewed (approved):
LGTM

vanwinkeljan reviewed (changes requested):
Please update the display shield `Kconfig.defconfig`s to check if LVGL is enabled or not before setting any LVGL related properties.

Further I see an issue with the various display/gui samples which will break because the display driver is no longer enabled if a shield is selected. One solution here could be to extend the board conf overlay infrastructure (`<path to sample>/boards/<board name>.conf`) to also support shields


Display/GUI samples that are using shields:
* samples/gui/lvgl
* samples/display/cfb_shell
* samples/display/cfb
* samples/display/ili9340

erwango commented:
@vanwinkeljan, thanks for feedback. I've fixed the `LVGL` points.

About:
> Further I see an issue with the various display/gui samples which will break because the display driver is no longer enabled if a shield is selected. One solution here could be to extend the board conf overlay infrastructure (<path to sample>/boards/<board name>.conf) to also support shields
 
I'm having a try putting the setting under `if DISPLAY` / `LVGL`
This would keep things working w/o adding multiple files. This is similar to what is done for connectivity. 
Ideally, this should work also for the boards used today in the samples you mentioned. So we might be also able to remove most of the `board.conf` files in these samples.

vanwinkeljan commented on boards/shields/adafruit_2_8_tft_touch_v2/Kconfig.defconfig:
```diff
@@ -0,0 +1,32 @@
+
+if DISPLAY
+
+config ILI9340
```
ILI9340 can only be enabled if SPI is enabled, so possibly the default value for SPI needs to be set to `y`

vanwinkeljan commented on boards/shields/ssd1306/Kconfig.defconfig:
```diff
@@ -0,0 +1,42 @@
+
+if DISPLAY
+
+config SSD1306
```
SSD1306 depends on I2C

vanwinkeljan commented on boards/shields/waveshare_epaper/Kconfig.defconfig:
```diff
@@ -0,0 +1,42 @@
+
+if DISPLAY
+
+config SSD16XX
```
SSD16XX depends on SPI

erwango commented:
@ulfalizer, can you review once more ?

jfischer-phytec-iot commented on boards/shields/waveshare_epaper/Kconfig.defconfig:
```diff
@@ -0,0 +1,45 @@
+	default 250 if SHIELD_WAVESHARE_EPAPER_GDEH0213B72
+
+config LVGL_VER_RES
+	default 128
```
Should be:
```
default 128 SHIELD_WAVESHARE_EPAPER_GDEH029A1
default 120 SHIELD_WAVESHARE_EPAPER_GDEH0213B1
default 120 SHIELD_WAVESHARE_EPAPER_GDEH0213B72
```

erwango replied:
Fixed, thanks

jfischer-no reviewed (changes requested):
Thanks, just one comment.

erwango commented:
@ulfalizer, when you have some time, a new review would be welcome

ulfalizer commented on boards/Kconfig:
```diff
@@ -23,6 +23,9 @@ source "$(BOARD_DIR)/Kconfig.board"
 
+# Parse shields references
+# Don't do it as a menuconfig, as shield selection is a CMake feature.
+rsource "shields/*/Kconfig.shield"
```
FWIW, I prefer `source` over `rsource` where `rsource` isn't needed, because it makes the path easier to grep for.

The globbing kinda breaks grepping already though.

ulfalizer commented on boards/shields/frdm_cr20a/Kconfig.defconfig:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2019, NXP
+# Copyright (c) 2019 Linaro Limited
+# SPDX-License-Identifier: Apache-2.0
+#
```
Inconsistent header. Make it

```
# Copyright (c) 2019, NXP
# Copyright (c) 2019 Linaro Limited
# SPDX-License-Identifier: Apache-2.0

...
```

ulfalizer commented on boards/shields/frdm_cr20a/boards/frdm_k64f.defconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2019 Linaro Limited
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Make it

```
# Copyright (c) 2019, NXP
# Copyright (c) 2019 Linaro Limited
# SPDX-License-Identifier: Apache-2.0
```

ulfalizer commented on boards/shields/waveshare_epaper/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2019 Linaro Limited
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Ditto

ulfalizer reviewed (approved):
Some nits.

Did you consider putting the shield stuff in the main board `Kconfig.defconfig` file by the way (e.g. after a `# Shield configuration` comment)? It's a bit confusing that we have so many Kconfig files.

ulfalizer commented:
Also wondering if `Kconfig.shield` could be merged with the main board `Kconfig` file, to cut down on things a bit there.

SebastianBoe commented:
`Also wondering if Kconfig.shield could be merged with the main board Kconfig file, to cut down on things a bit there.

`

Wouldn't that require O(boards*shields) snippets of shield code instead of O(shields) ?
